### PR TITLE
feat: Add type definitions for Sign In With Google JavaScript API

### PR DESCRIPTION
**Pull request by aronhoyer**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

---

## Note from author

Please note that as these type definitions are not for an existing NPM package, nor do we know the current version number of the library we're typing, the rules `dt-header` and `npm-naming` are disabled in `tslint.json`.

### CODE REVIEW

File: types/gsi/tslint.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "extends": "@definitelytyped/dtslint/dt.json",
+    "rules": {
+        "dt-header": false,
+        "npm-naming": false
```
Comment (sheetalkamat) [inline types/gsi/tslint.json:5]: Remove these rules and fix errors.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/gsi/tslint.json (hunk 1 above)
Comment (aronhoyer) [reply types/gsi/tslint.json:5]: thank you for adding a review. what do you suggest we add as a version number? i've been unable to find a version number in the api's official documentation.

**Comment by peterblazejewicz** [comment]
shouldn't this maybe named `google-signin`, as per identification token used on SF?
https://stackoverflow.com/questions/tagged/google-signin
(in iOS/Andrid, they are using 'GoogleSignin' moniker, e.g.: https://github.com/googlesamples/google-services/blob/master/ios/signin/SignInExample/AppDelegate.h#L19)

File: types/gsi/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
+// Type definitions for non-npm package gsi (Sign In With Google JavaScript API) 0.0
```
Comment (peterblazejewicz) [inline types/gsi/index.d.ts:1]:
```suggestion
// Type definitions for non-npm package Sign In With Google JavaScript API
```

File: types/gsi/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+     * window.onGoogleLibraryLoad = () => {
+     *   google.accounts.id.initialize({
+     *     // ...
+     *   })
+     *   google.accounts.id.prompt()
+     * }
+     * </script>
+     * ```
+     */
+    onGoogleLibraryLoad: () => void;
```
Comment (peterblazejewicz) [inline types/gsi/index.d.ts:270]: this one can be undefined, IMO declaration should allow to check (real life):
```ts
if (!window.onGoogleLibraryLoad) {...
```

**Comment by aronhoyer** [comment]
> shouldn't this maybe named `google-signin`

Google Sign In is a different library (I believe) and is getting deprecated. I chose `gsi` because that's what's in the source URI: https://accounts.google.com/gsi/client (as seen [here](https://developers.google.com/identity/gsi/web/guides/client-library))

File: types/gsi/index.d.ts (hunk 2 above)
Comment (aronhoyer) [reply types/gsi/index.d.ts:1]: removing version number will break the `dt-header` ts-lint rule

Comment (peterblazejewicz) [reply types/gsi/index.d.ts:1]: added back to suggestion, you're correct

File: types/gsi/index.d.ts (hunk 3 above)
Comment (aronhoyer) [reply types/gsi/index.d.ts:270]: good point. updated!

Comment (aronhoyer) [reply types/gsi/index.d.ts]: good point. updated!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Comment by aronhoyer** [comment]
Ready to merge

**Pull request merged**